modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+        // the name that comes from the enterprise downloader cluster state doesn't include the .mmdb extension,
+        // but the downloading and indexing of database code expects it to be there, so we add it on here before further processing
+        processDatabase(auth, name + ".mmdb", sha256, tgzUrl);
+    }
+
+    private void processDatabase(PasswordAuthentication auth, String name, String sha256, String url) {
+        Metadata metadata = state.getDatabases().getOrDefault(name, Metadata.EMPTY);
+        if (Objects.equals(metadata.sha256(), sha256)) {
+            updateTimestamp(name, metadata);
+            return;
+        }
+        logger.debug("downloading geoip database [{}]", name);
+        long start = System.currentTimeMillis();
+        try (InputStream is = httpClient.get(auth, url)) {
+            int firstChunk = metadata.lastChunk() + 1; // if there is no metadata, then Metadata.EMPTY + 1 = 0
+            Tuple<Integer, String> tuple = indexChunks(name, is, firstChunk, MessageDigests.sha256(), sha256, start);
+            int lastChunk = tuple.v1();
+            String md5 = tuple.v2();
+            if (lastChunk > firstChunk) {
+                state = state.put(name, new Metadata(start, firstChunk, lastChunk - 1, md5, start, sha256));
+                updateTaskState();
+                logger.info("successfully downloaded geoip database [{}]", name);
+                deleteOldChunks(name, firstChunk);
+            }
+        } catch (Exception e) {
+            logger.error(() -> "error downloading geoip database [" + name + "]", e);
+        }
+    }
+
+    // visible for testing
+    void deleteOldChunks(String name, int firstChunk) {
+        BoolQueryBuilder queryBuilder = new BoolQueryBuilder().filter(new MatchQueryBuilder("name", name))
+            .filter(new RangeQueryBuilder("chunk").to(firstChunk, false));
+        DeleteByQueryRequest request = new DeleteByQueryRequest();
+        request.indices(DATABASES_INDEX);
+        request.setQuery(queryBuilder);
+        client.execute(
+            DeleteByQueryAction.INSTANCE,
+            request,
+            ActionListener.wrap(r -> {}, e -> logger.warn("could not delete old chunks for geoip database [" + name + "]", e))
+        );
+    }
+
+    // visible for testing
+    protected void updateTimestamp(String name, Metadata old) {
+        logger.debug("geoip database [{}] is up to date, updated timestamp", name);
+        state = state.put(
+            name,
+            new Metadata(old.lastUpdate(), old.firstChunk(), old.lastChunk(), old.md5(), System.currentTimeMillis(), old.sha256())
+        );
+        updateTaskState();
+    }
+
+    void updateTaskState() {
+        PlainActionFuture<PersistentTask<?>> future = new PlainActionFuture<>();
+        updatePersistentTaskState(state, future);
+        state = ((EnterpriseGeoIpTaskState) future.actionGet().getState());
+    }
+
+    // visible for testing
+    Tuple<Integer, String> indexChunks(
```

Comment:
It may eventually be nice to re-use all the pieces from `GeoIpDownloader` rather than duplicate them. (Not now though, but at a later time). Maybe we could add that as a comment?

Reply:
Yeah I think the intention is to share a lot more between them (and possibly have them be the same thing). But given tight timelines we didn't want to accidentally break the widely-used GeoIpDownloader in this release.

Reply:
Yes, exactly this.